File: forthic-py/src/interpreter.py
import time
import operator
import pytz
import collections
from .tokens import (
    StringToken,
    CommentToken,
    StartArrayToken,
    EndArrayToken,
    StartModuleToken,
    EndModuleToken,
    StartDefinitionToken,
    EndDefinitionToken,
    StartMemoToken,
    WordToken,
    EOSToken,
    Token,
)
from .tokenizer import Tokenizer

from .module import Module, Word, PushValueWord, DefinitionWord
from .global_module import GlobalModule
from .profile import WordProfile
from .interfaces import IInterpreter, IModule, IWord
from typing import List, Any, Dict, Optional


# ----- Errors -----------------------------------------------------------------------------------------------
class InterpreterError(RuntimeError):
    pass


class UnknownModuleError(InterpreterError):
    def __init__(self, name: str):
        super().__init__(f"Can't find module: '{name}'")


class UnknownTokenError(InterpreterError):
    def __init__(self, token: Token):
        super().__init__(f'Unknown token: {token}')


class NestedDefinitionError(InterpreterError):
    def __init__(self):
        super().__init__("Can't have nested definitions")


class UnmatchedEndDefinitionError(InterpreterError):
    def __init__(self):
        super().__init__('Unmatched end definition')


class UnknownWordError(InterpreterError):
    def __init__(self, word_name: str):
        super().__init__(f"Unknown word: '{word_name}'")


# ----- Word Types -------------------------------------------------------------------------------------------
class EndArrayWord(Word):
    """This represents the end of an array"""
    def __init__(self):
        super().__init__(']')

    def execute(self, interp: IInterpreter) -> None:
        items: List[Any] = []
        item = interp.stack_pop()
        while not isinstance(item, StartArrayToken):
            items.append(item)
            item = interp.stack_pop()
        items.reverse()
        interp.stack_push(items)


class StartModuleWord(Word):
    """This indicates the start of a module

    See `docs/ARCHITECTURE.md` for more details on modules.
    """
    def __init__(self, name: str):
        super().__init__(name)

    def execute(self, interp: IInterpreter) -> None:
        # The app module is the only module with a blank name
        if self.name == '':
            interp.module_stack_push(interp.app_module)
            return

        # If the module is used by the current module, push it onto the module stack;
        # otherwise, create a new module and push that onto the module stack.
        module = interp.cur_module().find_module(self.name)

        # Check app module
        if not module:
            module = interp.app_module.find_module(self.name)

        if not module:
            module = Module(self.name, interp)
            interp.cur_module().register_module(module.name, module)

        interp.module_stack_push(module)


class EndModuleWord(Word):
    def __init__(self):
        super().__init__('}')

    def execute(self, interp: IInterpreter) -> None:
        interp.module_stack_pop()


class AppModule(Module):
    """The AppModule contains the words and variables of a Forthic application

    The app module is a speical module. This is the first module on the module stack. All applications start
    here. It is the only module where `USE-MODULE` can be called. It is the only nameless module.
    """
    def __init__(self, interp: 'Interpreter'):
        super().__init__('', interp)
        # Screens map names to chunks of Forthic code
        self.screens: Dict[str, str] = collections.defaultdict(str)

    def set_screen(self, name: str, content: str):
        self.screens[name] = content

    def get_screen(self, name: str) -> str:
        return self.screens[name]


class Interpreter(IInterpreter):
    """Interprets Forthic strings

    Modules may be registered with an Interpreter to provide more functionality.
    """
    def __init__(self, timezone=None):
        if not timezone:
            timezone = pytz.timezone('US/Pacific')
        self.timezone = timezone
        self.stack: List[Any] = []
        self.global_module = GlobalModule(self, self.timezone)
        self.app_module = AppModule(self)
        self.module_stack: List[IModule] = [self.app_module]
        self.registered_modules: Dict[str, Module] = {}
        self.is_compiling: bool = False
        self.is_memo_definition: bool = False
        self.cur_definition: Optional[DefinitionWord] = None
        self._dev_mode: bool = False

        # Profiling support
        self.word_counts: Dict[IWord, int] = collections.defaultdict(int)
        self.is_profiling: bool = False
        self.start_profile_time: Optional[float] = None
        self.timestamps: List[Any] = []
        self.cur_word_profile: WordProfile = None

    @property
    def dev_mode(self) -> bool:
        """This is used to indicate that things like debugging are ok"""
        return self._dev_mode

    @dev_mode.setter
    def dev_mode(self, dev_mode: bool):
        self._dev_mode = dev_mode

    def run(self, string: str) -> None:
        """Interprets a Forthic string, executing words one-at-a-time until the end of the string"""
        tokenizer = Tokenizer(string)
        token = tokenizer.next_token()
        while not isinstance(token, EOSToken):
            self.handle_token(token)
            token = tokenizer.next_token()

    def run_in_module(self, module: IModule, string: str) -> None:
        """Runs a Forthic string in the context of a given module"""
        self.module_stack_push(module)
        self.run(string)
        self.module_stack.pop()

    def cur_module(self) -> IModule:
        """The top of the module stack is the currently active module"""
        result = self.module_stack[-1]
        return result

    def find_module(self, name: str) -> Module:
        """Returns the module registered under the specified `name`"""
        if name not in self.registered_modules:
            raise UnknownModuleError(name)
        result = self.registered_modules[name]
        return result

    def stack_push(self, val: Any) -> None:
        """Pushes a value onto the Forth stack"""
        self.stack.append(val)

    def stack_pop(self) -> Any:
        """Pops a value from the Forth stack and returns it"""
        result = self.stack.pop()
        return result

    def module_stack_push(self, module: IModule) -> None:
        """Makes the specified `module` the active module"""
        self.module_stack.append(module)

    def module_stack_pop(self) -> IModule:
        """Removes the current module from the stack and makes the next module the current module"""
        return self.module_stack.pop()

    def register_module(self, module_class):
        """Registers an instance of Module with the interpreter

        Modules are typically registered at code time. This is where new capabilities can be made available
        to Forthic programs.
        """
        module = module_class(self)
        self.registered_modules[module.name] = module

    def find_word(self, name: str) -> Optional[IWord]:
        """Searches the interpreter for a word

        The module stack is searched top down. If the words cannot be found, the global module is searched.
        Note that the bottom of the module stack is always the application module.
        """
        modules = reversed(self.module_stack)
        result = None
        for m in modules:
            result = m.find_word(name)
            if result:
                break

        if not result:
            result = self.global_module.find_word(name)
        return result

    # --------------------------------------------------------------------------
    # Profiling support

    def start_profiling(self) -> None:
        """Clears word counts and starts profiling word executions"""
        self.is_profiling = True
        self.timestamps = []
        self.start_profile_time = time.perf_counter()
        self.add_timestamp('START')
        self.word_counts = collections.defaultdict(int)

    def add_timestamp(self, label: str) -> None:
        """Adds a timestamped label to a profiling run"""
        if not self.is_profiling or not self.start_profile_time:
            return
        self.timestamps.append(
            {
                'label': label,
                'time': time.perf_counter() - self.start_profile_time,
            }
        )

    def count_word(self, w: IWord) -> None:
        """If profiling, count word"""
        if self.is_profiling:
            self.word_counts[w] += 1

    def start_profile_word(self, word: IWord) -> None:
        """Used to mark the start of a word execution during a profiling run"""
        if not self.is_profiling:
            return

        word_profile = WordProfile(
            self.cur_word_profile, self.cur_module(), word
        )
        self.cur_word_profile = word_profile

    def end_profile_word(self) -> None:
        """Used to mark the end of a word execution during a profiling run"""
        if not self.cur_word_profile:
            return

        self.cur_word_profile.end_profile()
        parent = self.cur_word_profile.get_parent()
        if parent:
            self.cur_word_profile = parent

    def stop_profiling(self) -> None:
        """Stops profiling"""
        self.add_timestamp('END')
        self.is_profiling = False

    def word_histogram(self) -> List[Any]:
        """Returns a list of counts in descending order"""
        items = [
            {'word': w.name, 'count': c} for w, c in self.word_counts.items()
        ]
        result = sorted(items, key=operator.itemgetter('count'), reverse=True)
        return result

    def profile_timestamps(self) -> List[Any]:
        return self.timestamps

    # --------------------------------------------------------------------------
    # Handle tokens

    def handle_token(self, token: Token) -> None:
        """Called to handle each token from the Tokenizer"""
        if isinstance(token, StringToken):
            self.handle_string_token(token)
        elif isinstance(token, CommentToken):
            self.handle_comment_token(token)
        elif isinstance(token, StartArrayToken):
            self.handle_start_array_token(token)
        elif isinstance(token, EndArrayToken):
            self.handle_end_array_token(token)
        elif isinstance(token, StartModuleToken):
            self.handle_start_module_token(token)
        elif isinstance(token, EndModuleToken):
            self.handle_end_module_token(token)
        elif isinstance(token, StartDefinitionToken):
            self.handle_start_definition_token(token)
        elif isinstance(token, StartMemoToken):
            self.handle_start_memo_token(token)
        elif isinstance(token, EndDefinitionToken):
            self.handle_end_definition_token(token)
        elif isinstance(token, WordToken):
            self.handle_word_token(token)
        else:
            raise UnknownTokenError(token)

    def handle_string_token(self, token: StringToken) -> None:
        self.handle_word(PushValueWord('<string>', token.string))

    def handle_start_module_token(self, token: StartModuleToken) -> None:
        """Start/end module tokens are treated as IMMEDIATE words *and* are compiled"""
        word = StartModuleWord(token.name)
        if self.is_compiling:
            if not self.cur_definition:
                raise InterpreterError("Interpreter is compiling, but there is no current definition")
            self.cur_definition.add_word(word)

        # NOTE: We execute the word within a definition so we can do lookups during compile
        self.count_word(word)
        word.execute(self)

    def handle_end_module_token(self, token: EndModuleToken) -> None:
        word = EndModuleWord()
        if self.is_compiling:
            if not self.cur_definition:
                raise InterpreterError("Interpreter is compiling, but there is no current definition")
            self.cur_definition.add_word(word)

        # NOTE: We execute the word within a definition so we can do lookups during compile
        self.count_word(word)
        word.execute(self)

    def handle_start_array_token(self, token: StartArrayToken) -> None:
        self.handle_word(PushValueWord('<start_array_token>', token))

    def handle_end_array_token(self, token: EndArrayToken) -> None:
        self.handle_word(EndArrayWord())

    def handle_comment_token(self, token: CommentToken) -> None:
        pass

    def handle_start_definition_token(self, token: StartDefinitionToken) -> None:
        if self.is_compiling:
            raise NestedDefinitionError()
        self.cur_definition = DefinitionWord(token.name)
        self.is_compiling = True
        self.is_memo_definition = False

    def handle_start_memo_token(self, token: StartMemoToken) -> None:
        if self.is_compiling:
            raise NestedDefinitionError()
        self.cur_definition = DefinitionWord(token.name)
        self.is_compiling = True
        self.is_memo_definition = True

    def handle_end_definition_token(self, token: EndDefinitionToken) -> None:
        if not self.is_compiling:
            raise UnmatchedEndDefinitionError()
        if not self.cur_definition:
            raise InterpreterError("Cannot finish definition because no 'cur_definition'")

        if self.is_memo_definition:
            self.cur_module().add_memo_words(self.cur_definition)
        else:
            self.cur_module().add_word(self.cur_definition)
        self.is_compiling = False

    def handle_word_token(self, token: WordToken) -> None:
        word = self.find_word(token.name)
        if word is None:
            raise UnknownWordError(token.name)

        self.handle_word(word)

    def handle_word(self, word: IWord) -> None:
        if self.is_compiling:
            if not self.cur_definition:
                raise InterpreterError("Interpreter is compiling, but there is no current definition")
            self.cur_definition.add_word(word)
        else:
            self.count_word(word)
            word.execute(self)

File: experimental/forthic-swift/Forthic/Sources/Forthic/Interpreter.swift
import Foundation
enum InterpreterError: Error {
    case UNKNOWN_MODULE_ERROR(module_name: String)
    case UNKNOWN_TOKEN_ERROR(token: Token)
    case NESTED_DEFINITION_ERROR
    case UNMATCHED_END_DEFINITION_ERROR
    case UNKNOWN_WORD_ERROR(word_name: String)
    case STACK_UNDERFLOW
    case MODULE_STACK_UNDERFLOW
    case COMPILING_WITHOUT_CUR_DEFINITION
    case UNKNOWN_MODULE(module_name: String)
    case UNKNOWN_SCREEN(name: String)
}


/// Represents the end of an array
class EndArrayWord : Word {
    init() {
        super.init(name: "]")
    }
    
    override func execute(interp: Interpreter) throws {
        var items: [Any?] = []
        var item: Any? = try interp.stack_pop()
        while (!(item is StartArrayToken)) {
            items.append(item)
            item = try interp.stack_pop()
        }
        items.reverse()
        interp.stack_push(items)
    }
}


/// This represents a word that is defined in terms of other words
///
/// A definition looks like this:
///
/// : WORD-NAME   WORD1 WORD2 WORD3;
///
/// The name of the defined word is `WORD-NAME`. When it is executed, `WORD1`, `WORD2`, and `WORD3` are
/// executed in that order.
class DefinitionWord : Word {
    var words: [Word] = []

    func add_word(word: Word) {
        self.words.append(word)
    }
    
    override func execute(interp: Interpreter) throws {
        for w in self.words {
            try w.execute(interp: interp)
        }
    }
}

/// Pushes module onto interpreter's module stack
class StartModuleWord : Word {
    override init(name: String) {
        super.init(name: name)
    }
    
    override func execute(interp: Interpreter) throws {
        // The app module is the only module with a blank name
        if (self.name == "") {
            interp.module_stack_push(module: interp.app_module())
            return
        }
        
        // If the module is used by the current module, push it onto the module stack
        var module = interp.cur_module().find_module(name: self.name)
        
        // ..if not in the current module, check the app_module
        if (module == nil) {
            module = interp.app_module().find_module(name: self.name)
        }
        
        // ..otherwise, create a new module and push that onto the module stack
        if (module == nil) {
            let new_module = Module(name: self.name, interp: interp, forthic: "")
            interp.cur_module().register_module(module_name: new_module.name, module: new_module)
            module = new_module
        }
        
        interp.module_stack_push(module: module!)
    }
}

class EndModuleWord : Word {
    init() {
        super.init(name: "}")
    }
    
    override func execute(interp: Interpreter) throws {
        try interp.module_stack_pop()
    }
}


/// The application module is a special module that contains the words and variables for a Forthic application. There is only one
/// application module in an app. It's the first module on the module stack. It's the only module where `USE-MODULE` can be called.
/// It's the only module with no name.
///
/// The Forthic code for an application module may be split across several "screens" of code.
class AppModule: Module {
    var screens : [String: String] = [:]
    
    init(interp: Interpreter) {
        super.init(name: "", interp: interp, forthic: "")
    }
    
    func set_screen(name: String, forthic: String) {
        self.screens[name] = forthic
    }
    
    func get_screen(name: String) throws -> String {
        let result: String? = self.screens[name]
        if (result == nil) {
            throw InterpreterError.UNKNOWN_SCREEN(name: name)
        }
        return result!
    }
}

class Interpreter {
    var _app_module: AppModule? = nil
    var stack : [Any?] = []
    var _global_module : GlobalModule? = nil
    var module_stack: [Module] = []
    var registered_modules: [String: Module] = [:]
    var is_compiling: Bool = false
    var cur_definition: DefinitionWord? = nil
    var dev_mode: Bool = false
    var timezone: TimeZone
    
    init(timezone: TimeZone? = nil) {
        if (timezone == nil) {
            self.timezone = TimeZone.init(identifier: "America/Los_Angeles")!
        }
        else {
            self.timezone = timezone!
        }
        
        self._global_module = GlobalModule(interp: self, timezone: self.timezone)
        self._app_module = AppModule(interp: self)
        self.module_stack.append(self._app_module!)
    }
    
    func app_module() -> AppModule {
        return self._app_module!
    }

    func global_module() -> GlobalModule {
        return self._global_module!
    }

    /// Interprets a Forthic string, executing words one at a time until the end of the string
    func run(forthic: String) throws {
        let tokenizer = Tokenizer(string: forthic)
        var token = try tokenizer.next_token()
        while (!(token is EOSToken)) {
            try self.handle_token(token)
            token = try tokenizer.next_token()
        }
    }

    func run_in_module(module: Module, forthic: String) throws {
        self.module_stack_push(module: module)
        try self.run(forthic: forthic)
        try self.module_stack_pop()
    }

    func stack_push(_ value: Any?) {
        self.stack.append(value)
    }
    
    func stack_pop() throws -> Any? {
        if (self.stack.count == 0) {
            throw InterpreterError.STACK_UNDERFLOW
        }
        let result = self.stack.popLast()!
        return result
    }
    
    func cur_module() -> Module {
        let result = self.module_stack.last!
        return result
    }
    
    func find_module(name: String) throws -> Module {
        if (!self.registered_modules.keys.contains(name)) {
            throw InterpreterError.UNKNOWN_MODULE(module_name: name)
        }
        let result: Module = self.registered_modules[name]!
        return result
    }

    func module_stack_push(module: Module) {
        self.module_stack.append(module)
    }

    func module_stack_pop() throws {
        let module = self.module_stack.popLast()
        if (module == nil) {
            throw InterpreterError.MODULE_STACK_UNDERFLOW
        }
    }
    
    /// Registers a Module with the Interpreter so that it can be used from an application
    func register_module(module: Module) {
        self.registered_modules[module.name] = module
    }
    
    
    // ----- Private functions ----------------------------------------------------------------------------------------
    
    /// Searches the interpreter for a word
    ///
    /// The module stack is searched top down. If the words cannot be found, the global module is searched.
    /// Note that the bottom of the module stack is always the application module
    private func find_word(name: String) -> Word? {
        let modules = self.module_stack.reversed()
        var result: Word? = nil
        
        for m in modules {
            result = m.find_word(name: name)
            if (result != nil) {
                break
            }
        }
        
        if (result == nil) {
            result = self.global_module().find_word(name: name)
        }
        return result
    }
    
    /// Called to handle each token from the Tokenizer
    private func handle_token(_ token: Token) throws {
        if (token is StringToken) {
            try self.handle_string_token(token as! StringToken)
        }
        else if (token is CommentToken) {
            self.handle_comment_token(token as! CommentToken)
        }
        else if (token is StartArrayToken) {
            try self.handle_start_array_token(token as! StartArrayToken)
        }
        else if (token is EndArrayToken) {
            try self.handle_end_array_token(token as! EndArrayToken)
        }
        else if (token is StartModuleToken) {
            try self.handle_start_module_token(token as! StartModuleToken)
        }
        else if (token is EndModuleToken) {
            try self.handle_end_module_token(token as! EndModuleToken)
        }
        else if (token is StartDefinitionToken) {
            try self.handle_start_definition_token(token as! StartDefinitionToken)
        }
        else if (token is EndDefinitionToken) {
            try self.handle_end_definition_token(token as! EndDefinitionToken)
        }
        else if (token is WordToken) {
            try self.handle_word_token(token as! WordToken)
        }
        else {
            throw InterpreterError.UNKNOWN_TOKEN_ERROR(token: token)
        }
    }
    
    private func handle_string_token(_ token: StringToken) throws {
        try self.handle_word(PushValueWord(name: "<string>", value: token.str))
    }
    
    private func handle_comment_token(_ token: CommentToken) {
    }
    
    private func handle_start_array_token(_ token: StartArrayToken) throws {
        try self.handle_word(PushValueWord(name: "[", value: token))
    }
    
    private func handle_end_array_token(_ token: EndArrayToken) throws {
        try self.handle_word(EndArrayWord())
    }
    
    /// NOTE: This is treated as an IMMEDIATE word and is executed even if we're compiling
    private func handle_start_module_token(_ token: StartModuleToken) throws {
        let word = StartModuleWord(name: token.name)
        if (self.is_compiling) {
            if (self.cur_definition == nil) {
                throw InterpreterError.COMPILING_WITHOUT_CUR_DEFINITION
            }
            self.cur_definition?.add_word(word: word)
        }
        try word.execute(interp: self)
    }
    
    private func handle_end_module_token(_ token: EndModuleToken) throws {
        let word = EndModuleWord()
        if (self.is_compiling) {
            if (self.cur_definition == nil) {
                throw InterpreterError.COMPILING_WITHOUT_CUR_DEFINITION
            }
            self.cur_definition?.add_word(word: word)
        }
        try word.execute(interp: self)
    }
    
    private func handle_start_definition_token(_ token: StartDefinitionToken) throws {
        if (self.is_compiling) {
            throw InterpreterError.NESTED_DEFINITION_ERROR
        }
        self.cur_definition = DefinitionWord(name: token.name)
        self.is_compiling = true
    }
    
    private func handle_end_definition_token(_ token: EndDefinitionToken) throws {
        if (!self.is_compiling) {
            throw InterpreterError.UNMATCHED_END_DEFINITION_ERROR
        }
        if (self.cur_definition == nil) {
            throw InterpreterError.COMPILING_WITHOUT_CUR_DEFINITION
        }
        self.cur_module().add_word(word: self.cur_definition!)
        self.is_compiling = false
    }
    
    private func handle_word_token(_ token: WordToken) throws {
        let word = self.find_word(name: token.name)
        if (word == nil) {
            throw InterpreterError.UNKNOWN_WORD_ERROR(word_name: token.name)
        }
        try self.handle_word(word!)
    }

    private func handle_word(_ word: Word) throws {
        if (self.is_compiling) {
            if (self.cur_definition == nil) {
                throw InterpreterError.COMPILING_WITHOUT_CUR_DEFINITION
            }
            self.cur_definition?.add_word(word: word)
        }
        else {
            try word.execute(interp: self)
        }
    }
}
